AmbiguousMatchException: Multiple custom attributes of the same type found

From [this post](https://groups.google.com/d/msgid/exceldna/1b35daea-18b0-4b1f-917d-175a29266e38o%40googlegroups.com) in the mailing-list:

> 
> From: Johan A
> Date: Tue, Jul 7, 2020 at 9:39 AM
> Subject: [ExcelDna] Exception troubles with ExcelDna.Doc
> To: Excel-DNA
> 
> Hello,
> 
> I'm using ExcelDna in one of my Office Addin projects. I have upgraded to the latest versions of the needed ExcelDna modules
> 
> -ExcelDna 1.1.1
> -ExcelDna Integration 1.1.0
> -ExcelDna Intellisense 1.3.0
> -ExcelDna Interop 1.4.0.1
> -ExceDna Registration 1.1.0
> -ExcelDna Doc 0.2.3
> 
> The project is using .Net Framework 4.8 and building for AnyCpu.
> 
> However, I can't get the ExcelDna.Doc to build properbly. This is the output in final part of the build
> 
> ```
> creating HTML Help content in C:\VS\Source\Repos\MyProj\MyProj_ExcelDNA\bin\Debug\HelpContent
> 5>  ExcludeHidden: False, SkipCompile: False, Async: False
> 5>  
> 5>  using local template : ProjectFileTemplate.cshtml
> 5>  
> 5>  Unhandled Exception: System.Reflection.AmbiguousMatchException: Multiple custom attributes of the same type found.
> 5>     at System.Attribute.GetCustomAttribute(MemberInfo element, Type attributeType, Boolean inherit)
> 5>     at ExcelDnaDoc.Utility.ModelHelper.CreateFunctionModel(MethodInfo method, String defaultCategory)
> 5>     at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
> 5>     at System.Linq.Enumerable.<SelectManyIterator>d__17`2.MoveNext()
> 5>     at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
> 5>     at System.Linq.Lookup`2.Create[TSource](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
> 5>     at System.Linq.GroupedEnumerable`3.GetEnumerator()
> 5>     at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
> 5>     at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
> 5>     at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__1.MoveNext()
> 5>     at ExcelDna.Documentation.Models.AddInModel.<get_Functions>d__13.MoveNext()
> 5>     at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
> 5>     at CompiledRazorTemplates.Dynamic.dbadfeecadeade.Execute()
> 5>     at RazorEngine.Templating.TemplateBase.RazorEngine.Templating.ITemplate.Run(ExecuteContext context)
> 5>     at RazorEngine.Razor.Parse[T](String razorTemplate, T model)
> 5>     at ExcelDnaDoc.Templates.ViewBase`1.Publish()
> 5>     at ExcelDnaDoc.HtmlHelp.Create(String dnaPath, String helpSubfolder, Boolean excludeHidden, Boolean skipCompile, Boolean runAsync)
> 5>     at ExcelDnaDoc.Program.Main(String[] args)
> 5>C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(5258,5): error MSB3073: The command ""C:\VS\Source\Repos\packages\ExcelDnaDoc.1.1.0-beta2\tools\ExcelDnaDoc.exe" "C:\VS\Source\Repos\MyProj\MyProj_ExcelDNA\bin\Debug\MyProj_ExcelDNA-AddIn.dna" /Y" exited with code -532462766.
> ```
> 
> As you can see in the output, I have also tried to upgrade to the latest pre-release version of ExcelDnaDoc, but with same result.
> 
> Please, can someone give me a clue whats wrong. This has worked fine before.
> 
> Br,
> Johan
>

--- a/src/ExcelDnaDoc/Utility/ExcelDnaHelper.cs
+++ b/src/ExcelDnaDoc/Utility/ExcelDnaHelper.cs
@@ -30,8 +30,7 @@ namespace ExcelDnaDoc.Utility
 
             if (explicitExports)
             {
-                var excelFunction = (ExcelFunctionAttribute)Attribute.GetCustomAttribute(method, typeof(ExcelFunctionAttribute));
-                return (excelFunction != null);
+                return Attribute.IsDefined(method, typeof(ExcelFunctionAttribute));
             }
 
             ParameterInfo[] parameters = method.GetParameters();
@@ -45,8 +44,7 @@ namespace ExcelDnaDoc.Utility
                 return false;
             }
 
-            var excelCommand = (ExcelCommandAttribute)Attribute.GetCustomAttribute(method, typeof(ExcelCommandAttribute));
-            return (excelCommand != null);
+            return Attribute.IsDefined(method, typeof(ExcelCommandAttribute));
         }
     }
 }

--- a/src/ExcelDnaDoc/Utility/ModelHelper.cs
+++ b/src/ExcelDnaDoc/Utility/ModelHelper.cs
@@ -22,10 +22,12 @@
                 Category = defaultCategory
             };
 
-            var excelCommand = (ExcelCommandAttribute)Attribute.GetCustomAttribute(method, typeof(ExcelCommandAttribute));
+            var excelCommand = Attribute.GetCustomAttributes(method, typeof(ExcelCommandAttribute))
+                .OfType<ExcelCommandAttribute>()
+                .FirstOrDefault();
 
             // check if ExcelCommandAttribute used
-            if (excelCommand != null)
+            if (!(excelCommand is null))
             {
                 if (excelCommand.Name != null) { command.Name = excelCommand.Name; }
                 if (excelCommand.Description != null) { command.Description = excelCommand.Description; }
@@ -45,9 +47,11 @@
                 Description = string.Empty
             };
 
-            var excelArgument = (ExcelArgumentAttribute)Attribute.GetCustomAttribute(parameter, typeof(ExcelArgumentAttribute));
+            var excelArgument = Attribute.GetCustomAttributes(parameter, typeof(ExcelArgumentAttribute))
+                .OfType<ExcelArgumentAttribute>()
+                .FirstOrDefault();
 
-            if (excelArgument != null)
+            if (!(excelArgument is null))
             {
                 if (excelArgument.Name != null) { model.Name = excelArgument.Name; }
                 if (excelArgument.Description != null) { model.Description = excelArgument.Description; }
@@ -76,11 +80,13 @@
                     IsHidden = false
                 };
 
-            var excelFunction = (ExcelFunctionAttribute)Attribute.GetCustomAttribute(method, typeof(ExcelFunctionAttribute));
-            var excelFunctionSummary = (ExcelFunctionDocAttribute)Attribute.GetCustomAttribute(method, typeof(ExcelFunctionDocAttribute));
+            var customAttributes = Attribute.GetCustomAttributes(method, typeof(ExcelFunctionAttribute));
+
+            var excelFunction = customAttributes.OfType<ExcelFunctionAttribute>().FirstOrDefault();
+            var excelFunctionSummary = customAttributes.OfType<ExcelFunctionDocAttribute>().FirstOrDefault();
 
             // check if ExcelFunctionDocAttribute used
-            if (excelFunctionSummary != null)
+            if (!(excelFunctionSummary is null))
             {
                 if (excelFunctionSummary.Summary != null) { function.Summary = excelFunctionSummary.Summary; }
                 if (excelFunctionSummary.Returns != null) { function.Returns = excelFunctionSummary.Returns; }
@@ -91,10 +97,9 @@
             {
                 //Use reflection to check for matching properties or fields
                 //Adds support to allow people to use their own classes that inherit from ExcelFunctionAttribute
-                var attribs = Attribute.GetCustomAttributes(method);
-                foreach (var attrib in attribs)
+                foreach (var attrib in customAttributes)
                 {
-                    if (attrib != null && typeof(ExcelFunctionAttribute).IsAssignableFrom(attrib.GetType()))
+                    if (!(attrib is null))
                     {
                         string summary = (string)attrib.GetType().GetProperty("Summary")?.GetValue(attrib, null);
                         if (string.IsNullOrEmpty(summary))
@@ -121,7 +126,7 @@
             }
 
             // check if ExcelFunctionAttribute used
-            if (excelFunction != null)
+            if (!(excelFunction is null))
             {
                 if (excelFunction.Name != null) { function.Name = excelFunction.Name; }
                 if (excelFunction.Description != null) { function.Description = excelFunction.Description; }

